### ISSUE TITLE
very slow with TrajectoryTrainer

### ISSUE THREAD

**Issue by GengSS** [mir-group/flare#144]
Hi,

Thank you for the FLARE,

I am learning to use FLARE with simple examples,
here is an example:
md.traj is a md trajectory of Ni16 cluster,  with EAM Potential, totally 5000 frames
```
from flare.gp import GaussianProcess
from flare.mc_simple import two_plus_three_body_mc, two_plus_three_body_mc_grad

gp = GaussianProcess(kernel=two_plus_three_body_mc, kernel_grad=two_plus_three_body_mc_grad,
                     hyps=[0.01, 0.01, 0.01, 0.01, 0.01],
                     cutoffs = (7,7),
                     hyp_labels=['Two-Body Signal Variance','Two-Body Length Scale','Three-Body Signal Variance',
                                'Three-Body Length Scale', 'Noise Variance'])

from flare.struc import Structure
from ase.io import read
from ase.calculators.singlepoint import SinglePointCalculator

structures = []
images = read("./create_database_with_eam/md.traj",index="::25",format="traj")

for a in images:
    e = a.get_potential_energy()
    f = a.get_forces()
    b = a.copy()
    b.set_cell([25,25,25])
    b.set_calculator(SinglePointCalculator(b,energy=e,forces=f))
    structures.append(Structure.from_ase_atoms(b))
    if len(structures) > 100:
        break

print("length of structures are {}".format(len(structures)))
from flare.gp_from_aimd import TrajectoryTrainer
TT = TrajectoryTrainer(frames=structures,
                    gp = gp,
                    rel_std_tolerance = 3,
                    abs_std_tolerance=0,
                    pre_train_on_skips=5,
                    verbose=2,
                    checkpoint_interval=10,
                    n_cpus=4)

TT.run()
print("Done!")
```

I ran this script, with 4 CPUs. 
However, after printing a line 'Train GP' in the file of gp_from_aimd.out,
it seems to get stuck and never prints anything.

Is there anything wrong in the script?
Thank you very much.

Best Regards,

Geng

**Comment by stevetorr** [comment]
Hi Geng,

Thank you for _using_ FLARE! I'm sorry you are having issues.
There could be a few things behind this. One way to start debugging would be to ensure that FLARE is working properly on your system by running unit tests.

First, make sure you have pytest installed on your python installation, and then navigate to flare/tests. You can run:

`pytest test_gp.py' and `pytest test_gp_from_aimd.py'.
If both of these hang then it means there is not something wrong with your script, but something wrong with the way FLARE is working on your computer.

Also, if you are using pre_train_on_skips = 5 and have 5000 frames total, then you'll be adding in 16 atoms from each frame, and using every fifth, so 1000 frames * 16 atoms = 16000 atoms into your training set. This is a **HUGE** Gaussian Process, so the fact it'll take forever is not a surprise :) I would suggest using a much, much more conservative pre_train_on_skips value. Honestly, if your system is simple, you could get away with 1250 or even 2500 to start; just dump in four frames =64 atoms or two =32 to start with and see how the trajectory trainer handles it from there.

You could also remove the pre_train_on_skips argument and use the argument `seed_frames=structures[:1]', which will just use the first frame to start with.

Also, are you running on a cluster? That can also introduce it's own problems, but, first why don't you try the above and let me know :)

**Comment by GengSS** [comment]
Dear Steven,

Thank you very much for your reply.
I used the pytest to check whether installation worked properly or not.
There are the results:
`==================================== test session starts =====================================
platform linux -- Python 3.6.10, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /u/scratch/g/gengsun/flare_research/flare
collected 9 items                                                                            

test_gp.py .........                                                                   [100%]

================================ 9 passed in 71.14s (0:01:11) ================================


==================================== test session starts =====================================
platform linux -- Python 3.6.10, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /u/scratch/g/gengsun/flare_research/flare
collected 4 items                                                                                                          

test_gp_from_aimd.py ....                                                                                            [100%]

==================================================== 4 passed in 35.67s ====================================================
`

Both of them finished correctly within reasonable walltimes. So I suppose the FLARE installation is correct.

Then I went back to my test example. At this time, I only use 10 images (hence 160 atoms) and 4 CPUs. I found that 'gp_from_aimd-hyps.dat' gets updated every 15 min, like this:
`
GP hyperparameters: 
Hyp0 : 0.01
Hyp1 : 0.01
Hyp2 : 0.01
Hyp3 : 0.01
Hyp4 : 0.01
likelihood: -4472.9497894923015
likelihood gradient: [-5.46789336e-02  5.50601874e-02 -4.32851149e+04  4.28596782e+04
 -9.61891004e+03]

GP hyperparameters: 
Hyp0 : 0.009999766703801536
Hyp1 : 0.010001284661705872
Hyp2 : 0.009999766703801536
Hyp3 : 1.0099999999990934
Hyp4 : 0.009999766703801536
likelihood: -1185.7147172754585
likelihood gradient: [-36404.86052155  80761.04930398  -6895.08031257    324.2452749
  -9501.50021536]

`
I wonder is this a reasonable walltime as the code is expected? Normally, what could be a reasonable number of atoms used for FLARE? I would like to get some experience as a baseline.
Thank you very much.

Best Regards,

Geng

**Comment by stevetorr** [comment]
Hey Geng,

Thanks for your patience! Typically we run it on 32 cores on our local supercomputer cluster, and so, I don't know exactly what your local circumstances are (if the 4 cores are on your laptop or what).

@nw13slx has put in a lot of labor into parallelization via multiprocessing.  She recently benchmarked, for this issue, a run on 188 atomic environments on 16 cores: it took 6 seconds for one call of `set_L_alpha` which then means for optimizing the gradient it will take take 6 seconds times the number of hyperparameters (3 for 2-body and 5 for 2+3 body). This was also on the development branch, which has some speedups over the master branch. So that'll take maybe 30 minutes? to optimize hyperparameters in a good case. 

The hyperparameter optimzation is one of the more time-consuming parts of training, until your set gets huge. After around 200-400 atoms in your training set you start to gain huge from multiprocessing. 

I can also benchmark a little on my own and let you know. How have you been finding things in the last few days?
